### ISSUE TITLE
improvements to Correct_3D_drift.py

### ISSUE THREAD

**Issue by tischi** [fiji/fiji#136]
Hello!

Encouraged by Albert Cardona I changed (hopefully improved) the code of Correct_3D_drift.py in quite some places. In fact, I did not do this "for fun" but we have various data that required all these changes as the current version  of Correct_3D_drift would not work as is.

Changed code:
- also works now for __2D time-series__ (used to be 3D only)
- added option: __measure drift on multiple timescales__. This allows to also find slow drift components of less than 1 pixel per frame (used to not be detected).
- added option: __correct sub-pixel drift__ computing the shifted images using the Translate() function of TransformJ (used to do integer shifts only)
- added option: compute drift on __edge filtered__ images (for some of my data the underlying "PhaseCorrelation" algorithm finds the drift better in edge filtered images than in the raw data)
- added option: if a __ROI__ is put on the image, only this part of the image is considered for drift computation; the ROI is moved along with the detected drift thereby __tracking__ the structure of interest and keeping it stable in the corrected time-series.

Basically I was wondering if I could upload my code somewhere so that you could check whether it might be worth to become part of Fiji, possibly even replacing the old script?

By the way, is this the current script? https://github.com/fiji/fiji/blob/master/plugins/Scripts/Plugins/Registration/Correct_3D_drift.py
I was confused, because at least in my Windows Fiji line 16 reads:
`from javax.vecmath import Point3i`
whereas in above github it reads:
`from org.scijava.vecmath import Point3i`

Looking forward to your answers! 
Christian

**Comment by imagejan** [comment]
> I was confused, because at least in my Windows Fiji line 16 reads:
`from javax.vecmath import Point3i`
whereas in above github it reads:
`from org.scijava.vecmath import Point3i`

Indeed, this was changed by @ctrueden in the process of migrating [ImageJ to Java 8](http://imagej.net/2015-12-22_-_The_road_to_Java_8), but the fix (https://github.com/fiji/fiji/commit/d0d02ce8d9bce59c4b2db03cfaaf4ca597701d23) didn't make it to the Fiji update site yet.

> Basically I was wondering if I could upload my code somewhere

The best way to [contribute](http://imagej.net/Contributing) is by submitting a pull request. You can simply click on the small pencil icon on the top right on [this page](https://github.com/fiji/fiji/blob/master/plugins/Scripts/Plugins/Registration/Correct_3D_drift.py), which will create a fork of this repository and lets you edit the script online, so you can paste your code and open the pull request.

**Comment by tischi** [comment]
Thanks for the clarification and the recipe for contribution, which I followed.

**Comment by imagejan** [comment]
@tischi For some reason your [PR](https://github.com/tischi/fiji/pull/1) ended up on your own fork of the `fiji` repository, instead of the original one. Could you please try again and make sure that `fiji/fiji` is selected as the *base fork*?

Regarding your changes, I noticed that you introduce a dependency to `imagescience`. The imagescience library recently moved to its own update site due to licensing issues, and is not any more included in Fiji, so this might make updating this script difficult, because users need to have the ImageScience update site enabled for the script to run as expected.

Maybe @ctrueden or @hinerm can comment on this (when they're back from travelling).

**Comment by tischi** [comment]
@imagejan I did another pull request stating:
`tischi  wants to merge 1 commit into fiji:master from tischi:tischi-patch-1`
is this correct now?

@ctrueden @hinerm 
regarding the __imagescience dependency__:
I need a function for sub-pixel translation of a 3D data set. This works well:
https://github.com/imagescience/ImageScience/blob/master/src/main/java/imagescience/transform/Translate.java
However, if you know of another function that is easier in terms of dependencies I could use that instead. In fact, the imagescience implementation has the other drawback of __not supporting multi-core__, making it rather slow for big data sets. I feel it should be possible to implement translations in 3D using linear interpolation with multi-core support, however I also feel that this is outside my programming comfort zone. But maybe __imglib2__ has something already?

**Comment by ctrueden** [comment]
Closed by #137.

Regarding the imagescience dependency: it would indeed be best to eliminate that. Otherwise, the script will not work in plain Fiji anymore. The `imglib2-realtransform` library has a [RealViews](https://github.com/imglib/imglib2-realtransform/blob/imglib2-realtransform-2.0.0-beta-30/src/main/java/net/imglib2/realtransform/RealViews.java) class that supports subpixel translation, but of course it operates on ImgLib2 data structures. You can wrap an `ImagePlus` into an ImgLib2 image using the `wrap` methods of [ImageJFunctions](https://github.com/imglib/imglib2-ij/blob/imglib2-ij-2.0.0-beta-31/src/main/java/net/imglib2/img/display/imagej/ImageJFunctions.java) in `imglib2-ij`. Maybe this is good enough?

**Comment by tischi** [comment]
I will check whether I can replace the imagescience code with imglib2 code!

**Comment by tischi** [comment]
@ctrueden 
Kota and me tried several hours to use imglib2, but so far failed. Basically, it seems that the __wrap__ function does not convert the ImagePlus enough to be used by the imglib2 algorithms.
I paste the code below. Can you help?

    # import
    from net.imglib2.img.display.imagej import ImageJFunctions
    from net.imglib2.img import ImagePlusAdapter
    from net.imglib2.realtransform import RealViews, Translation3D
    from ij import IJ
    from script.imglib.algorithm import Affine3D
    from net.imglib2.img import Img, ImagePlusAdapter
    import jarray
    from io.scif.img import ImgOpener
    from net.imglib2.img.array import ArrayImgFactory
    from net.imglib2.type.numeric.integer import UnsignedByteType
    from ij import ImagePlus

    # get the image
    imp = IJ.getImage()
    img = ImagePlusAdapter.wrapReal(imp)
    print img

    # translation
    dx=10.5; dy=5; dz=0;
 
    # try translation
    trans3D = Translation3D(dx, dy, dz)
    img_translated = RealViews.transformReal(img, trans3D)

    # try affine transform
    m = [1,0,0,dx,\
         0,1,0,dy,\
         0,0,1,dz]
    jm = jarray.array(m, 'f')
    img_affine = Affine3D(img, jm, Affine3D.LINEAR)

**Comment by ctrueden** [comment]
I'll try to find time to dig into this...

**Comment by tischi** [comment]
@ctrueden Hi Curtis, __you don't need to__! We already made quite some progress with the help of Stephan Saalfeld. Right now we are in e-mail conversation with Stephan because to me it looks a bit strange what imglib2 is doing in some cases using below code. Here is code that works in principle:

    from net.imglib2.img.display.imagej import ImageJFunctions
    from net.imglib2.img import ImagePlusAdapter
    from net.imglib2.realtransform import RealViews, Translation3D
    from net.imglib2.img import Img, ImagePlusAdapter
    from net.imglib2.img.array import ArrayImgFactory
    from net.imglib2.type.numeric.integer import UnsignedByteType
    from net.imglib2.view import Views
    from net.imglib2.img.imageplus import ImagePlusImgs
    from net.imglib2.interpolation.randomaccess import NLinearInterpolatorFactory
    from ij import IJ, ImagePlus

    # get image
    imp = IJ.getImage()
    
    # translation using imglib2
    img = ImagePlusImgs.from(imp) 
    # prepare image from translation using interpolation 
    extended = Views.extendBorder(img) 
    interpolant = Views.interpolate(extended, NLinearInterpolatorFactory()) 
    # translation
    dx=10.5; dy=5; dz=0.5;
    transform = Translation3D(dx, dy, dz)
    transformed = RealViews.affine(interpolant, transform)
    cropped = Views.interval(transformed, img)
    imp_transformed = ImageJFunctions.wrap(cropped,"imglib2")
    imp_transformed.show()

**Comment by ctrueden** [comment]
I'm so glad you guys are on it! And thanks @axtimwalde for helping get this to work! :hatched_chick:

**Comment by axtimwalde** [comment]
@tischi you see an overflow of n-linear interpolation due to accuracy limits of unsigned bytes.  The generic NLinearInterpolator on NumericType uses the source type to do accumulation.  That can lead to overflow in types with limited precision.  We just had a discussion about the lack of precision guarantees and write hierarchies in ImgLib2 types https://github.com/imglib/imglib2/pull/120#issuecomment-179913135 and you just hit it.  So what can you do?  There is an ARGB version of the NLinearInterpolator that explicitly clamps but you would use that only for ARGBColor images.  For single channel images, you would do the work in float or double precision and later coy back.  converting the imageplus is that fastest option but has a memory footprint.  Instead, you could

    [...]
    extended = Views.extendBorder(img)
    converted = Converters.convert(extended, RealFloatConverter())
    interpolant = Views.interpolate(extended, NLinearInterpolatorFactory())
    [...]

but only for singe channel, ARGBInterpolator for the rest, you will find that on your own.

**Comment by imagejan** [comment]
> `interpolant = Views.interpolate(extended, NLinearInterpolatorFactory())`

I guess you meant:

```
interpolant = Views.interpolate(converted, NLinearInterpolatorFactory())
                                ^^^^^^^^^
```
Right?
Thanks for the enlightening discussion!

**Comment by axtimwalde** [comment]
You're absolutely right. Written in a rush and not tested nor proofread...

**Comment by tischi** [comment]
`img = ImagePlusImgs.from(imp.duplicate())`
`extended = Views.extendBorder(img)`
`converted = Converters.convert(extended, RealFloatConverter())`

gives me the following error: 
    converted = Converters.convert(extended, RealFloatConverter())
TypeError: convert(): 2nd arg can't be coerced to net.imglib2.converter.readwrite.SamplerConverter,

I think the convert method needs in this case a 3rd argument, which is called B b in the javadoc and, frankly I did not understand what that should be...
http://javadoc.imagej.net/ImgLib2/net/imglib2/converter/Converters.html#convert(net.imglib2.IterableInterval,%20net.imglib2.converter.Converter,%20B)

could you help again?

**Comment by cmci** [comment]
Hi Tisch,

Maybe use RealDoubleSamplerConverter which is an implementation of
SamplerConverter?

onverted = Converters.convert(extended, RealFloatSamplerConverter())

see
http://javadoc.imagej.net/ImgLib2/net/imglib2/converter/readwrite/SamplerConverter.html

Cheers,
Kota (from Rohrbach)

On Mon, Feb 8, 2016 at 11:21 AM, tischi <notifications@github.com> wrote:

> img = ImagePlusImgs.from(imp.duplicate())
> extended = Views.extendBorder(img)
> converted = Converters.convert(extended, RealFloatConverter())
>
> gives me the following error:
> converted = Converters.convert(extended, RealFloatConverter())
> TypeError: convert(): 2nd arg can't be coerced to
> net.imglib2.converter.readwrite.SamplerConverter,
>
> I think the convert method needs in this case a 3rd argument, which is
> called B b in the javadoc and, frankly I did not understand what that
> should be...
>
> http://javadoc.imagej.net/ImgLib2/net/imglib2/converter/Converters.html#convert(net.imglib2.IterableInterval,%20net.imglib2.converter.Converter,%20B)
>
> could you help again?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fiji/fiji/issues/136#issuecomment-181294898>.
>



-- 

-------------------------------------------------------------
Dr. Kota Miura <miura@embl.de>

Senior Image Analyst
European Molecular Biology Laboratory
Meyerhofstr. 1, 69117 Heidelberg, GERMANY

Tel +49 6221 387 8404
Mobile +49 160 95001177http://cmci.embl.de
++++++++++++++++++

Dr. Kota Miura  <miura@nibb.ac.jp <miura@nins.jp>>
Research Fellow
National Institute of Basic Biology

Okazaki, 444-8585
JAPAN
-------------------------------------------------------------

**Comment by StephanPreibisch** [comment]
Exactly, thank Kota!

		Img img = ImagePlusImgs.from(imp.duplicate());
		RandomAccessible extended = Views.extendBorder(img);
		RandomAccessible< FloatType > converted = Converters.convert(extended, new RealFloatSamplerConverter());

**Comment by tischi** [comment]
Thanks Kota! I have something working now. The wrapping back to imp is not very beautiful due to the explicit type casting, is it possible to avoid this?

```
from net.imglib2.img.display.imagej import ImageJFunctions
from net.imglib2.realtransform import RealViews, Translation3D
from net.imglib2.view import Views
from net.imglib2.img.imageplus import ImagePlusImgs
from net.imglib2.converter import Converters
from net.imglib2.converter.readwrite import RealFloatSamplerConverter
from net.imglib2.interpolation.randomaccess import NLinearInterpolatorFactory
from ij import IJ, ImagePlus

def translate3D_singleStack_imglib2(imp, dx, dy, dz):
  '''
  3D translation using imglib2
  '''
  ## wrap
  img = ImagePlusImgs.from(imp.duplicate())
  ## prepare image
  extended = Views.extendBorder(img)
  converted = Converters.convert(extended, RealFloatSamplerConverter())
  interpolant = Views.interpolate(converted, NLinearInterpolatorFactory())
  ## translate
  transform = Translation3D(dx, dy, dz)
  transformed = RealViews.affine(interpolant, transform)
  cropped = Views.interval(transformed, img)
  ## wrap back and return
  bd = imp.getBitDepth()
  if bd==8:
    return(ImageJFunctions.wrapUnsignedByte(cropped,"imglib2"))
  elif bd == 16:
    return(ImageJFunctions.wrapUnsignedShort(cropped,"imglib2"))
  elif bd == 32:
    return(ImageJFunctions.wrapFloat(cropped,"imglib2"))
  else:
    return None    
  

imp = IJ.getImage()
imp_transformed = translate3D_singleStack_imglib2(imp, dx=10, dy=0.5, dz=0.5)
imp_transformed.show()
```

**Comment by axtimwalde** [comment]
@tischi @StephanPreibisch @cmci This is not a good 'fix' because it uses a writable converter where you need read access only, limiting the application to writable source types for no good reason.  A short glimpse into [Javadoc](http://javadoc.imagej.net/ImgLib2/net/imglib2/converter/RealFloatConverter.html) and [source code](https://github.com/imglib/imglib2/blob/master/src/main/java/net/imglib2/converter/Converters.java#L75-L83) reveals that `B`, for `RealFloatConverter`, is `FloatType`, so

    from net.imglib2.type.numeric.real import FloatType
    from net.imglib2.img.display.imagej import ImageJFunctions
    from net.imglib2.realtransform import RealViews, Translation3D
    from net.imglib2.view import Views
    from net.imglib2.img.imageplus import ImagePlusImgs
    from net.imglib2.converter import Converters, RealFloatConverter
    from net.imglib2.interpolation.randomaccess import NLinearInterpolatorFactory
    from ij import IJ, ImagePlus

    imp = IJ.getImage()
    img = ImagePlusImgs.from(imp.duplicate())
    extended = Views.extendBorder(img)
    converted = Converters.convert(extended, RealFloatConverter(), FloatType())

Sorry for the sloppy response earlier, I really didn't test (and you guys really didn't read the docs ;)).

Cheers,
Stephan

**Comment by axtimwalde** [comment]
And expanding on this a bit because it is educational: Why does the method need B as a parameter?  Because the Converter source and target types are completely generic.  `RealFloatConverter` is a bad teaching example because it makes a commitment about the target.  `Converter<A,B>` does not.  In this general case, the `convert` method has to make sure that there is a `Type` template that can be used to generate `B` target types for accessors.

**Comment by cmci** [comment]
I stepped on the generic programming paradigm with my muddy feet... thank you for your clear advice. 

So the question now relates to the last question of Tischi: could original `imp` object be somehow used as an argument for the the wrapper for reverting the image to ImagePlus so that we don't need three different cases of bit depth?

By the way, I had some chances to see codes in ImgLib1 and witnessed your (and @StephanPreibisch ) tremendous efforts in many case operations for different bit-depth to go back to ImagePlus. Great works, no wonder being generic @axtimwalde !

**Comment by StephanPreibisch** [comment]
Hi Stephan,

I didn’t really check if there is a difference in behaviour between the different static convert() methods, was quick as usual ;) It is obvious when looking at the source code though.

Maybe it would be good if RW/RO properties of the SamplerConverter vs Converter is reflected in the static method or interface names so it is more obvious that there is an important difference? Right now it is basically only clear from the package names. What do you think?

Cheers,
Stephan

**Comment by tischi** [comment]
Hi All,

It would be great if you could help me with another question. How could I modify above code to also work when the input **imp** is an image with **two channels**? Above code crashes during the wrapping back with an ArrayIndexOutOfBounds exception. I don't know whether the RealViews.affine command actually does the right thing but it is running also when imp has two channels.

**Comment by axtimwalde** [comment]
Hi @tischi, check out this http://javadoc.imagej.net/ImgLib2/net/imglib2/view/Views.html#collapseReal%28net.imglib2.RandomAccessible,%20int%29 , you want to collapse the channels dimension into a vector, as long as this is a vector of numeric or real types, interpolation works as expected.

**Comment by tischi** [comment]
Hi @axtimwalde: thanks a lot!

I think (hope) I found the right spot in the conversion flow to add the **collapseReal()** (see below).
However, now I am stuck wrapping the **cropped** img back to an multi-channel imp. I had a look here: http://javadoc.imagej.net/ImgLib2/net/imglib2/img/display/imagej/ImageJFunctions.html
But I only found functions working on single channel images, or did I overlook something?!


```
img = ImagePlusImgs.from(imp.duplicate())
extended = Views.extendBorder(img)
converted = Converters.convert(extended, RealFloatSamplerConverter())
composite = Views.collapseReal(converted, imp.getNChannels())
interpolant = Views.interpolate(composite, NLinearInterpolatorFactory())
transformed = RealViews.affine(interpolant, Translation3D(dx, dy, dz))
cropped = Views.interval(transformed, img)
```

**Comment by axtimwalde** [comment]
Yes @tischi, that function is missing.  Can you have a look at the code and add it?  Pull requests more than welcome ;).  I think it would be a great idea to add this.

**Comment by tischi** [comment]
Ok @axtimwalde, fair enough! 

However, I had a look at the Java code of the ImageJFunctions and I am afraid that this is way beyond my understanding of imglib2 (which is close to zero) and in addition, I usually only use jython and hardly ever write code in Java. I think realistically I would need someone here with a bit of insight into imglib2 in order to teach me how implement things like this.

I guess, in terms of my programming skills I have to resort to splitting the channels in ImageJ and processing them individually and combining them again afterwards.

**Comment by axtimwalde** [comment]
> I guess, in terms of my programming skills I have to resort to splitting the channels in ImageJ and processing them individually and combining them again afterwards.

@tischi sure, that certainly works, for splitting, have a look at http://javadoc.imagej.net/ImgLib2/net/imglib2/view/Views.html#hyperSlice%28net.imglib2.RandomAccessible,%20int,%20long%29 

Regarding ImageJFunctions code complexity, I think you overestimate it.  As a start, you could try to simply copy it into an ImagePlusImg.  The vrtual stack approach isn't currently compatible with composite types.

**Comment by tpietzsch** [comment]
One way to make it work with minimal coding would be:
* make a `Converter` from component `d` of the `Composite<T>` to `T` (didn't look at the actual types)
* `Converters.convert` to make `T` views for each of the composite components
* `Views.stack` these
* then use `ImageJFunctions` on the result.

**Comment by axtimwalde** [comment]
@tpietzsch true, that is a good solution.  @tischi, the only thing you would have to write is a

```java
CompositeChannelConverter<T> implements Converter<Composite<T>, T>
```

because that is the only piece that shouldn't be done from Jython as it will be called per pixel access.

**Comment by axtimwalde** [comment]
Hi @tischi, I added this converter https://github.com/imglib/imglib2/commit/82cd11b333299ba8e7147d405cf298fea6cb4d2c, we haven't made a release though, so to use it you would have to install a recent snapshot of ImgLib2 http://maven.imagej.net/service/local/artifact/maven/redirect?r=snapshots&g=net.imglib2&a=imglib2&v=2.7.0-SNAPSHOT&e=jar

**Comment by tischi** [comment]
Hi @axtimwalde, thanks a lot! I will try it asap!

**Comment by tischi** [comment]
Is there a deadline from which on `imagescience` will not be shipped with Fiji anymore?

**Comment by hinerm** [comment]
@tischi I believe they have already been split out? But all that changed is that they stopped shipping with the **core** (Fijij) update site.

@emeijering maintains the ImageScience [update site](http://imagej.net/ImageScience) and I don't see any reason these plugins would cease to be available there

**Comment by imagejan** [comment]
> I believe they have already been split out?

@hinerm currently the imagescience libraries are still served by *both* the Fiji and the ImageScience update sites (the latter one overriding the former).

From the [announcement](http://imagej.net/2016-01-05_-_ImageScience_is_moving_to_an_update_site):
> We will leave the current versions of the ImageScience plugins on the Fiji update site for a few more weeks, to give people time to transition to the update site.

So there wasn't a deadline announced yet for removal from the Fiji update site.

**Comment by emeijering** [comment]
@tischi @hinerm The [ImageScience](http://imagej.net/ImageScience) library and plugin suites will indeed remain available ([source](https://github.com/imagescience) and [compiled](http://sites.imagej.net/ImageScience/)) so by all means feel free to continue using them.
